File: Advance/3.Write a program to check if a linked list is a palindrome/Solution.java
import java.util.*;

class Node {
    int data;
    Node next;
    Node(int val) {
        data = val;
        next = null;
    }
}

public class PalindromeLinkedList {
    public static void append(Node head, int data) {
        Node newNode = new Node(data);
        Node temp = head;
        while (temp.next != null) temp = temp.next;
        temp.next = newNode;
    }

    public static boolean isPalindrome(Node head) {
        Stack<Integer> stack = new Stack<>();
        Node temp = head;

        while (temp != null) {
            stack.push(temp.data);
            temp = temp.next;
        }

        temp = head;
        while (temp != null) {
            if (temp.data != stack.pop()) return false;
            temp = temp.next;
        }
        return true;
    }

    public static void main(String[] args) {
        Scanner scanner = new Scanner(System.in);
        System.out.print("Enter number of elements: ");
        int n = scanner.nextInt();

        System.out.print("Enter elements:\n");
        int val = scanner.nextInt();
        Node head = new Node(val);

        for (int i = 1; i < n; i++) {
            val = scanner.nextInt();
            append(head, val);
        }

        if (isPalindrome(head)) {
            System.out.println("The linked list is a palindrome.");
        } else {
            System.out.println("The linked list is not a palindrome.");
        }
        scanner.close();
    }
}

File: Advance/3.Write a program to check if a linked list is a palindrome/Solution.cpp
#include <iostream>
#include <stack>
using namespace std;

struct Node {
    int data;
    Node* next;
    Node(int val) : data(val), next(nullptr) {}
};

void append(Node*& head, int data) {
    if (!head) {
        head = new Node(data);
        return;
    }
    Node* temp = head;
    while (temp->next) temp = temp->next;
    temp->next = new Node(data);
}

bool isPalindrome(Node* head) {
    stack<int> st;
    Node* temp = head;

    while (temp) {
        st.push(temp->data);
        temp = temp->next;
    }

    temp = head;
    while (temp) {
        if (temp->data != st.top()) return false;
        st.pop();
        temp = temp->next;
    }
    return true;
}

int main() {
    Node* head = nullptr;
    int n, val;

    cout << "Enter number of elements: ";
    cin >> n;

    cout << "Enter elements:\n";
    for (int i = 0; i < n; i++) {
        cin >> val;
        append(head, val);
    }

    if (isPalindrome(head)) {
        cout << "The linked list is a palindrome." << endl;
    } else {
        cout << "The linked list is not a palindrome." << endl;
    }

    return 0;
}
